Use per-client persistent allocators on JITServer

dmitry-ten opened the pull request:
As described in eclipse#11162, a regular persistent allocator doesn't
release freed memory back to the system, which causes the memory
occupied by persistent allocator to remain unavailable for other
allocation types, even if all the clients have disconnected.
If the available memory is low, this leads to memory
exhaustion once a large enough number of  clients have cycled through.

This commit enables the server to create a new persistent allocator
for each client. When a client disconnects and its client data is
cleaned, its corresponding persistent allocator will also be destroyed.
This releases all memory reserved by that allocator back into the system.
We modify the persistent allocator to use virtual memory allocation,
instead of simple malloc, because memory allocated with malloc is not
guaranteed to be released back into the system and could also be
retained internally.

The main interface for using per-client allocation is inside
`TR::CompilationInfoPerThreadBase` and contains the following methods:
- `enterPerClientAllocationRegion(clientUID)`: begins per-client
persistent allocation on the current thread for a client with a given UID.
All allocations/frees that would normally be done globally, are now
rerouted to be done by a per-client allocator, e.g. `PERSISTENT_NEW`,
`TR::Compiler->persistentAllocator()`, `jitPersistentAlloc/Free`, etc.
- `exitPerClientAllocationRegion()`: end per-client allocation on the
current thread. All allocations/frees after this call will be done by
the default global allocator.

While the vast majority of persistent allocations can be strictly
divided into per-client/global allocations, there may be some rare
cases when a data structure is allocated globally but can be freed or
reallocated in the middle of a compilation. These cases need to be
handled manually by exiting the allocation region, performing the
required
memory operation and re-entering the region.

dmitry-ten commented:
Dependent on eclipse/omr#5760.

mpirvu reviewed (commented):
This is a risky change. In order to better check for errors I propose that, every time we need to allocate a persistent block we store the address of the allocator in the _next field of the block header
```
   // Persistent block header
   //
   struct Block
      {
      size_t _size;
      Block * _next;

      explicit Block(size_t size, Block * next = 0) : _size(size), _next(next) {}
      size_t size() const { return _size; }
      void setSize(size_t sz) { _size = sz; }
      Block * next() const { return _next; }
      void setNext(Block *b) { _next = b; }
      };
```
Thus, when we free we can check that we use the correct allocator.
I don't have a full-proof way of finding all the places where we need to switch back to the global persistent allocator. Maybe we don't really have to do so during compilation (I saw that you covered IdiomRecognition, but there are other places as well).
This switching back and forth between different allocators may be affected by exceptions as well. 

Other considerations:
We need a global option to turn this new mechanism off. 
How can we be absolutely sure that the two hashatbles, the one for storing <clientUID, clientSession> mappings and the one that stores <clientUID, allocator> are in sync?
If I remember correctly, the list of large freed blocks is protected by a monitor that is not included in the PersistentAllocator. Can we do something about this to avoid contention among clients?

dmitry-ten commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -858,6 +868,7 @@ ClientSessionHT::purgeOldDataIfNeeded()
                (unsigned long long)iter->first, (long long)oldAge);
             ClientSessionData::destroy(iter->second); // delete the client data
             _clientSessionMap.erase(iter); // delete the mapping from the hashtable
+            deletePerClientAllocator(iter->first); // delete the per-client allocator
```
It's not actually, need to move this up one line.

dmitry-ten commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -624,10 +633,14 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
+      // because there is a global pool of plans
+      exitPerClientAllocationRegion();
       // Build my entry
       if (!(optPlan = TR_OptimizationPlan::alloc(clientOptPlan.getOptLevel())))
          throw std::bad_alloc();
       optPlan->clone(&clientOptPlan);
+      enterPerClientAllocationRegion(clientId);
```
All the exceptions thrown here will set `abortCompilation` to true. In the part that handles that case, the allocation region is always exited.

dmitry-ten commented on runtime/compiler/optimizer/IdiomRecognition.cpp:
```diff
@@ -1205,6 +1211,12 @@ TR_CISCGraph::makePreparedCISCGraphs(TR::Compilation *c)
+#if defined(J9VM_OPT_JITSERVER)
+   // Return to per-client allocation
+   if (c->isOutOfProcessCompilation())
+      c->fej9()->_compInfoPT->enterPerClientAllocationRegion(c->fej9()->_compInfoPT->getClientData()->getClientUID());
```
So if the exception is thrown during the execution of `TR::CompilationInfPerThreadBase::compile`, there is potentially one region where we might be using incorrect allocator - the code executed after exception is thrown and before we return to global allocator some time after `compile` returns. I think the solution would be to re-enter per-client allocation in the catch block of `compile`.

dmitry-ten commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -794,7 +802,9 @@ ClientSessionHT::deleteClientSession(uint64_t clientUID, bool forDeletion)
          ClientSessionData::destroy(clientData); // delete the client data
+
          _clientSessionMap.erase(clientDataIt); // delete the mapping from the hashtable
+         deletePerClientAllocator(clientUID); // delete the per-client allocator
```
Could you clarify this one? I'm not sure I understand what order is required here.

mpirvu replied:
I was referring to the fact that this function is called with the compQueueMonitor in hand, and in here we acquire `_allocatorTableMonitor`. So the order of locking these two monitors must be obeyed everywhere else.

dmitry-ten commented:
I addressed most of the concerns in the review and followed the advice to store pointer to allocator used during `allocate` inside `next` field of `Block`.
I added the ability to disable per-client allocation by passing env variable `TR_DisablePerClientPersistentAllocation` to the server.

Looking at the usage of persistent memory during compilation, the only place I see that will cause problems is JProfiler, since we create profiling entries during compilation, but their lifetimes are manager by `JProfilerThread`. Since we don't support JProfiler yet it's not exactly clear to me which parts need to use global allocator, until I see the JITServer impelmentation.

> How can we be absolutely sure that the two hashtables, the one for storing <clientUID, clientSession> mappings and the one that stores <clientUID, allocator> are in sync?

If that's a major concern, we can store it in the same table like <clientUID, std::pair<clientSession, clientUID>>. I think it should be fine though, since we delete from both tables almost at the same time.

> If I remember correctly, the list of large freed blocks is protected by a monitor that is not included in the PersistentAllocator. Can we do something about this to avoid contention among clients?

Yes, I see that large blocks use the global `memoryAllocMonitor`. I don't see a reason why we can't just get rid of the global monitor and put it inside the allocator class. I think it's better to do it in a separate PR though, because this change would affect non-JITServer code as well.

AlexeyKhrabrov commented:
I've looked at this PR since I'm working on some dependent changes (#11777), and I have some questions/comments. Please correct me if my understanding is wrong.

- I think that the life cycle management of per-client allocators could be simplified. The allocator instance for a new client can be created in `ClientSessionHT::findOrCreateClientSession()`, and a pointer to it can be stored in `ClientSessionData`. The client session would basically own the allocator instance. `ClientSessionData` objects themselves can still be allocated using the per-client allocators. Maintaining a separate mapping from client IDs to their allocator would not be necessary, as well as the synchronization associated with creating allocator instances "on-demand" separately from client sessions. A pointer to the client session is already cached in `compInfoPT` after the server receives a compilation request but prior to making any persistent allocations, so the allocator instance can be accessed through the cached client session pointer.

- I think that in some of the code paths, e.g. in `CompilationInfoPerThreadRemote::processEntry()`, it would be better (more robust w.r.t. exception handling) to use the global or per-client allocator explicitly rather than rely on per-client vs. global allocation regions. Allocation regions should only be used in cases when using the global allocator explicitly would require a lot of changes to the code.

- `compInfoPT` or the client session might be readily available in some contexts, so the per-client allocator can be accessed without going through thread local storage. Although I'm not sure if avoiding the performance penalty is worth the effort to identify such cases.

dmitry-ten commented:
> think that the life cycle management of per-client allocators could be simplified. The allocator instance for a new client can be created in ClientSessionHT::findOrCreateClientSession(), and a pointer to it can be stored in ClientSessionData

Good point, I will do that

> it would be better (more robust w.r.t. exception handling) to use the global or per-client allocator explicitly rather than rely on per-client vs. global allocation regions. 

Originally I wasn't going to have a way to explicitly get the global allocator, and resolve all exceptions with regions, but then I realized that in some places that's the only way to get the correct allocator, so I had to add `persistentGlobalAllocator`. I guess you are right that it's better to allocate some things  like `ServerStream` explicitly using the global allocator.

> compInfoPT or the client session might be readily available in some contexts, so the per-client allocator can be accessed without going through thread local storage. Although I'm not sure if avoiding the performance penalty is worth the effort to identify such cases.

I do not like the idea of explicitly modifying code to use per-client allocator after going through the effort of making it automatic. So, unless, there is a noticeable performance penalty for accessing the thread-local `compInfoPT` I would rather leave things as is.

dmitry-ten commented:
I modified this PR to address @AlexeyKhrabrov's concerns and I also rewrote it to use per-client instances of `TR_PersistentMemory` in addition to allocators. This allows to do this change without making methods of `PersistentAllocator` virtual in OMR.

mpirvu commented on runtime/compiler/env/J9CompilerEnv.cpp:
```diff
@@ -66,3 +69,39 @@ J9::CompilerEnv::isCodeTossed()
+TR::PersistentAllocator &
+J9::CompilerEnv::persistentAllocator()
+   {
+#if defined(J9VM_OPT_JITSERVER)
```
As I mentioned in https://github.com/eclipse/omr/pull/5760 we should add a simple test to detect JITServer and eliminate TLS access overhead for non-jitserver.
The simplest approach is to make `PersistentInfo::_remoteCompilationMode` static because we already use it, it is set very early and never changes after it is set.

dmitry-ten replied:
Yes, I'll do it.
This made me realize that I need to deal with `compInfo->persistentMemory()` as well, since currently it always returns the global memory.
JITServer only retrieves persistent memory from compInfo in a couple of places, so I'm thinking to just replace them with calls to `TR::Compiler->persistentMemory()` and leave the `compInfo` memory alone, as it's also used to get persistent info and we want that to remain global.

dmitry-ten commented:
I fixed bugs and verified that this change actually works with many clients.
With #11823 included, I can run `sanity.functional` tests locally and the server does not crash.
So this should be ready for another round of review @mpirvu. 

One important change I had to made: since `TR_PersistentMemory` is also per-client now, and it's part of the `TR_Memory` object stored inside `Compilation`, calling `comp->trMemory()` needs to return the correct memory pointer, depending on whether we are inside or outside of per-client allocation region. To avoid further changes to OMR, I instead changed `J9::Compilation` object and added methods that replace its `_trMemory` pointer with either global or per-client one when we enter/exit per-client allocaiton region. This has less overhead than always calling `TR::Compiler->persistentMemory` from `TR_Memory` and avoids OMR changes but I feel that it's kind of an awkward solution. 

Another concern is that `::trPersistentMemory` and `compInfo->persistentMemory()`  still always return the global memory. The former because it's a global pointer so we can't change it for different clients and the latter because it is needed to access global `persistentInfo` and I also wanted to leave at least one way to access global memory if necessary that doesn't involve going through the global pointer.

In terms of memory performance I think the results are pretty good. In my experiment, I ran 20 AcmeAir clients with load  simultaneously, with server allocated 1GB of RAM. Then I shut down all the clients and repeated the load again.

In the first run:
- 700MB on the server after all JMeters finish
- 108MB after all clients shut down

In the second run:
- 716MB on the server after all JMeters finish
- 114MB after all clients shut down 

While there is a small increase between the first and second runs, this shows that the vast majority of persistent memory
gets freed after all clients disconnect. Without this change, the memory usage on the server does not go down at all.

mpirvu reviewed (commented):
I have some small questions/suggestion. Please see the inline comments

dmitry-ten commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -609,25 +615,29 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
 
+      // Optimization plan needs to use the global allocator,
+      // because there is a global pool of plans
+      exitPerClientAllocationRegion();
```
Right but the problem is that optimization plan is defined in OMR so we can't modify its allocator.

dmitry-ten commented on runtime/compiler/env/J9PersistentInfo.hpp:
```diff
@@ -321,8 +320,9 @@ class PersistentInfo : public OMR::PersistentInfoConnector
 #if defined(J9VM_OPT_JITSERVER)
+   static JITServer::RemoteCompilationModes _remoteCompilationMode; // JITServer::NONE, JITServer::CLIENT, JITServer::SERVER
+
    JITServer::RemoteCompilationModes getRemoteCompilationMode() const { return _remoteCompilationMode; }
```
I was under the impression that static methods can't be called on an instance so I left this as-is to avoid changing every call to this function. But apparently there is no such rule so I'll change it.

mpirvu commented:
@AlexeyKhrabrov  I would appreciate another review from you since you had some suggestions.
This item caries a bit of risk, so I would like it to be well scrutinized.
Also, we may need to do more testing than usual for it. @dmitry-ten do we need to deliver the supporting OMR change first?

dmitry-ten commented:
Yes, OMR change needs to be merged first.
I tested this with `sanity.functional` but yeah, running more test suits is in order.
And the tests have to be done locally too, since Jenkins runs will succeed even if the server crashes.

dmitry-ten commented:
Rebased to the latest master

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
jenkins test sanity all jdk11

AlexeyKhrabrov reviewed (commented):
Most of my comments have been addressed. I still have concerns about exception handling (see inline comments). Getting rid of TLS accesses could be done separately if performance impact is non-negligible (although I doubt that).

dmitry-ten commented:
I left a machine to do repeated runs of `_sanity.functional` and `_extended.functional` overnight, and I detected a bug:
`purgeOldDataIfNeeded` that's called by the statistics thread, will delete unused client sessions. It needs to enter a per-client allocation region before deleting the session. 
Another potential issue is here:
https://github.com/eclipse/openj9/blob/88975a47b8c3e8335fa6c24d25a7f48852b6b513/runtime/compiler/control/JITServerCompilationThread.cpp#L863
When compilation gets aborted early, we need to re-enter per-client allocation region as well, before deleting the client session.
I'll push a fix in a moment.

dmitry-ten commented on runtime/compiler/optimizer/IdiomRecognition.cpp:
```diff
@@ -1205,6 +1211,12 @@ TR_CISCGraph::makePreparedCISCGraphs(TR::Compilation *c)
          minimumHotnessPrepared = hotness;
       }
+
+#if defined(J9VM_OPT_JITSERVER)
+   // Return to per-client allocation
+   if (c->isOutOfProcessCompilation())
+      c->fej9()->_compInfoPT->enterPerClientAllocationRegion();
+#endif
```
I could swear I had to code to handle that.
Regardless, the solution should be to check whether we are in the per-client region in the catch block of `TR::CompilationInfoPerThread::wrappedCompile`  and re-enter it if not.

mpirvu replied:
Another solution would be to implement some RAII code (just like our critical section code) where the constructor of an object that is created on the stack performs `_compInfoPT->exitPerClientAllocationRegion()` and the destructor, executed when the stack is unwound, does the opposite calling enterPerClientAllocationRegion().

dmitry-ten commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -907,14 +937,17 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
    compInfo->updateCompilationErrorStats((TR_CompilationErrorCode)entry._compErrCode);
 
+
    TR_OptimizationPlan::freeOptimizationPlan(entry._optimizationPlan); // we no longer need the optimization plan
```
That's correct. I don't even know why this doesn't crash immediately.

mpirvu replied:
Strange, the code that I fetch in my local repo for review has `exitPerClientAllocationRegion();` at this very place

mpirvu replied:
`exitPerClientAllocationRegion();` from below needs to be moved up. I think it's worth checking why tests pass.

dmitry-ten commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -7133,6 +7134,31 @@ TR::CompilationInfoPerThreadBase::shouldPerformLocalComp(const TR_MethodToBeComp
+void
+TR::CompilationInfoPerThreadBase::enterPerClientAllocationRegion()
+   {
+   OMR::CriticalSection compilationMonitorLock(_compInfo.getCompilationMonitor());
```
You are right, it's not needed. The critical section is a leftover code from when I implemented this by accessing `ClientSessionHT` object directly. I'll remove it.

AlexeyKhrabrov commented:
> ... It needs to enter a per-client allocation region before deleting the session.
> ...
> When compilation gets aborted early, we need to re-enter per-client allocation region as well, before deleting the client session.

To avoid such cases, maybe `deleteClientSession()` should have an assertion that we're currently inside the correct per-client region. Same applies to `clearCaches()`.

Alternatively, all the data in the client session could be allocated and deallocated using its own persistent allocator explicitly. I think this would be more clean, but it would require more changes to the code, so I'm not sure if it's worth it.

AlexeyKhrabrov commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -7133,6 +7134,31 @@ TR::CompilationInfoPerThreadBase::shouldPerformLocalComp(const TR_MethodToBeComp
+   {
+   OMR::CriticalSection compilationMonitorLock(_compInfo.getCompilationMonitor());
+   TR_ASSERT(_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER, "must be called on the server"); 
+   static const char* disablePerClientPersistentAllocation = feGetEnv("TR_DisablePerClientPersistentAllocation");
+   if (!disablePerClientPersistentAllocation)
```
I would expect this option to have global effect and disable per-client allocation altogether, not just the regions. Was this the intention?

dmitry-ten replied:
Yes, the intention is to completely disable per-client allocation.

dmitry-ten commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -609,25 +615,29 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
+      exitPerClientAllocationRegion();
       // Build my entry
       if (!(optPlan = TR_OptimizationPlan::alloc(clientOptPlan.getOptLevel())))
          throw std::bad_alloc();
```
Yeah, we should re-enter the per-compilation region for deallocating client options and recompilation info in the catch block

dmitry-ten commented:
> Alternatively, all the data in the client session could be allocated and deallocated using its own persistent allocator explicitly. I think this would be more clean, but it would require more changes to the code, so I'm not sure if it's worth it.

Actually, thinking about how to deallocate a client session from inside the statistics thread, this might be the only option.
Because `TR::Compiler->persistentMemory` only works when we have the correct `compInfoPT` which we don't have in the statistics thread. It will mostly be easy to implement, the only complication should be the `JITServerPersistentCHTable`, because it will allocate subclasses outside of the client session code and we need to explicitly use the per-client allocator there as well.

dmitry-ten commented:
@mpirvu do you have any opinion on this?
I wanted to avoid having  to explicitly use per-client allocator as much as possible but I don't see another way to handle purging
client sessions when we don't have access to `compInfoPT`

mpirvu commented:
Indeed, if the deallocations can be done by a thread different that the one that performed the allocations, we must indicate explicitly what persistent memory to use.
Most of the changes can be contained within the JITServer code. I see an issue with
`TR_PersistentClassInfo::removeSubclasses()` which probably needs a parameter indicating the correct persistent memory. We can use a default parameter of NULL to avoid changing code in mainline.

> only complication should be the JITServerPersistentCHTable, because it will allocate subclasses outside of the client session code and we need to explicitly use the per-client allocator there as well.

I didn't follow the code, but why would the JITServerPersistentCHTable allocate anything outside the client session? I would assume that allocations are done by a compInfoperThread which has the correct allocator.

dmitry-ten commented:
@mpirvu I just pushed the latest changes to this PR.
Mainly, I changed `ClientSessionData` to explicitly use per-client memory when destroying the session. I left allocation code as-is because it is functionally correct, since we are guaranteed to be in the per-client region during allocation and my experience trying to explicitly use per-client allocation leads me to believe that it might make code more confusing (see below for my struggles).

I also modified code to disable per-client allocation to store the option in the client session so that we check the environment variable only  once.

As a side note, I determined the reason for weird bugs that I was investigating for the last few days. I will put the summary here, just in case.
The bug I was encountering was when using `operator new` and passing `TR_PersistentMemory *` to it, as in:
```
_vmInfo = new (_persistentMemory) VMInfo();
```
The operator wouldn't actually allocate the memory and, instead, it returned the address of `_persistentMemory` itself.
I was confused because I thought that including `TR_PERSISTENT_ALLOC(TR_Memory::ClientSessionData)` in the definition of client session would allow me to use `operator new` that takes `TR_PersistentMemory *` as a parameter. My mistake was not understanding that overloading `operator new` in a class, only overloads it when creating the instances of the class itself.
It does not enable me to use the overloaded operator when using `new` to allocate objects within the class methods. So what the code above was actually calling was the  `<new>` library function `void* operator new  ( std::size_t count, void* ptr );`, where `ptr` points to already allocated memory and just returns it unmodified. That's why it took me a while to realize what was actually happening. 
The solution to this is to declare `TR_PERSISTENT_ALLOC` for every class that we want to allocate using `operator new` with persistent memory. But I decided to keep `new (PERSISTENT_NEW)` for allocation of those classes as that seems like an easier solution that doesn't require adding new memory tags to OMR

dmitry-ten commented:
Fixed copyrights.

mpirvu commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -780,19 +791,24 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
+      enterPerClientAllocationRegion();
       if (clientOptions)
          {
          TR_Memory::jitPersistentFree(clientOptions);
```
Maybe we should specify the persistent memory to be used here as we know it's the one stored in the session:
`getClientData()->persistentMemory()->freePersistentMemory(clientOptions)`
`getClientData()` must exist if clientOptions have been created

dmitry-ten replied:
Done.

mpirvu commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -780,19 +791,24 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
+
       if (_recompilationMethodInfo)
          {
          TR_Memory::jitPersistentFree(_recompilationMethodInfo);
```
Similar here: `getClientData()->persistentMemory()->freePersistentMemory(_recompilationMethodInfo)`

dmitry-ten replied:
Done

mpirvu commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -780,19 +791,24 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
 
+      if (optPlan)
+         {
+         TR_OptimizationPlan::freeOptimizationPlan(optPlan);
```
Does `freeOptimizationPlan()` know how to get to global allocator?

dmitry-ten replied:
Yeah, it needs to be moved outside per-client region.
I think this bug is not caught by tests because the only way we can reach the catch block with allocated optimization plan is if its allocation throws, which shouldn't happen under normal memory conditions. All other exceptions should throw before the plan is allocated so we never actually reach this code during testing.

mpirvu commented on runtime/compiler/runtime/RuntimeAssumptions.hpp:
```diff
@@ -197,6 +198,7 @@ class TR_PersistentClassInfo : public TR_Link0<TR_PersistentClassInfo>
    flags8_t                            _flags;
    flags8_t                            _shouldNotBeNewlyExtended; // one bit for each possible compilation thread
    CCVResult                           _ccvResult;
+   TR_PersistentMemory                *_persistentMemory;
```
Adding an extra pointer to `TR_PersistentClassInfo` will increase footprint because we have such an entry for every class. It is the reason we only allow 8 compilation threads (`_shouldNotBeNewlyExtended` has a bit for every compilation thread and we decided to limit that at 8)

mpirvu replied:
My suggestion was to add an optional TR_Memory parameter to `removeSubClasses()` so that it knows exactly what persistent memory to use for deallocation. If the parameter is not passed, then it can use the global persistent memory. Interestingly, I don't see who calls `removeSubClasses()` for non-jitserver case.

dmitry-ten replied:
I see, I didn't think think footprint would be a concern here.
I'll modify it to do what you suggest.

AlexeyKhrabrov replied:
This is outside of the scope of this PR, but I think is worth noting:

If the footprint of this class is indeed a concern (8 bytes per class doesn't seem like a lot...), I think it can actually be reduced. The `_nameLength` field doesn't seem to be used anywhere, and removing it would reduce the size by 8 bytes (same as would be added by the `_persistentMemory` pointer). The last 3 fields could be reordered to allow extending `_shouldNotBeNewlyExtended` to `flags16_t`, which would then enable up to 16 compilation threads instead of 8 (assuming this is the only place that causes the limit). `_ccvResult` could also be encoded in the `_flags` if we ever need another 8 bits.

mpirvu replied:
Indeed. I didn't realize that `_nameLength` is no longer used. I have no recollection for what could have been used.
Maybe will do that in another PR.

mpirvu commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -7886,7 +7908,10 @@ TR::CompilationInfoPerThreadBase::compile(J9VMThread * vmThread,
       TR::Region dispatchRegion(regionSegmentProvider, rawAllocator);
-      TR_Memory trMemory(*_compInfo.persistentMemory(), dispatchRegion);
+
+      // Initialize trMemory with per-client persistent memory, since
```
Per-client persistent memory is only used for JITServer, so let's adjust the comment to reflect that

dmitry-ten replied:
Done.

mpirvu commented on runtime/compiler/env/JITServerPersistentCHTable.cpp:
```diff
@@ -46,7 +46,7 @@ JITServerPersistentCHTable::~JITServerPersistentCHTable()
       {
```
Above in the constructor of `JITServerPersistentCHTable` I am thinking that we could be more explicit in 
` _classMap(decltype(_classMap)::allocator_type(TR::Compiler->persistentAllocator()))`
and use the allocator that comes with the persistentMemory given as a parameter. I assume that persistentMemory is the per-client one.

dmitry-ten replied:
Done

dmitry-ten commented:
Addressed the comments

dmitry-ten commented:
Ok, there is an interaction between `requeueOutOfProcessEntry` and `freeOptimizationPlan` because the former puts the entry back into the pool, while the latter needs a valid entry to find the plan. So `freeOptimizationPlan` needs to save a pointer to the plan before recycling the entry. I'll need to change that.

mpirvu commented:
@dmitry-ten I didn't fully understand your last message, but looking at `requeueOutOfProcessEntry()` I see it calling `recycleCompilationEntry()` which calls `entry->freeJITServerAllocations()` which frees `_clientOptions` which is allocated with per-client persistent memory.
Maybe we should hang `_clientOptions` to the compInfoPerThreadRemote rather than MethodToBeCompiled and eliminate that `freeJITServerAllocations`. If we go this route we should create another PR that is delivered before this one.

dmitry-ten commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -816,8 +826,20 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
       setMethodBeingCompiled(NULL); // Must have the compQmonitor
 
+
+      // Save the pointer to the plan before recycling the entry
+      TR_OptimizationPlan *optPlan = entry._optimizationPlan;
+      // This deallocates client options, so need to use a per-client allocator
       compInfo->requeueOutOfProcessEntry(&entry);
 
+      exitPerClientAllocationRegion();
+      if (optPlan)
+         {
+         TR_OptimizationPlan::freeOptimizationPlan(optPlan);
+         }
```
I meant this is what I need to do to correctly free the optimization plan.
Moving client options outside of `MethodToBeCompiled` sounds like a better idea though

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
jenkins test sanity all jdk11

mpirvu commented:
All tests have passed.
@AlexeyKhrabrov are you ok with current implementation?

dmitry-ten commented on runtime/compiler/compile/J9Compilation.cpp:
```diff
@@ -185,6 +185,8 @@ J9::Compilation::Compilation(int32_t id,
    _remoteCompilation(false),
    _serializedRuntimeAssumptions(getTypedAllocator<SerializedRuntimeAssumption *>(self()->allocator())),
    _clientData(NULL),
+   _globalMemory(TR_Memory(*::trPersistentMemory, heapMemoryRegion)),
```
Done

dmitry-ten commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -8808,6 +8833,8 @@ TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary *portLib, void *
       if (compiler && compiler->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
          {
          compiler->getOptions()->closeLogFileForClientOptions();
+         if (!that->getPerClientPersistentMemory()) // threw while outside of per-client region, re-enter it here
```
Yes, I would rather put it in a separate PR since this is already a large change and has been in progress for a long time.

mpirvu commented:
jenkins test sanity all jdk11

mpirvu commented:
All tests have passed, hence merging.
